server/sandbox_run_linux.go

```diff
@@ -37,6 +39,234 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/types"
 )
 
+// addToMappingsIfMissing ensures the specified id is mapped from the host.
+func addToMappingsIfMissing(ids []idtools.IDMap, id int64) []idtools.IDMap {
+	firstAvailable := int(0)
+	for _, r := range ids {
+		if int(id) >= r.HostID && int(id) < r.HostID+r.Size {
+			// Already present, nothing to do
+			return ids
+		}
+		if r.ContainerID+r.Size > firstAvailable {
+			firstAvailable = r.ContainerID + r.Size
+		}
+	}
+	newMapping := idtools.IDMap{
+		ContainerID: firstAvailable,
+		HostID:      int(id),
+		Size:        1,
+	}
+	return append(ids, newMapping)
+}
+
+func (s *Server) configureSandboxIDMappings(mode string, sc *pb.LinuxSandboxSecurityContext) (*storage.IDMappingOptions, error) {
+	if mode == "" {
+		// No mode specified but mappings set in the config file, let's use them.
+		if s.defaultIDMappings != nil {
+			uids := s.defaultIDMappings.UIDs()
+			gids := s.defaultIDMappings.GIDs()
+			return &storage.IDMappingOptions{UIDMap: uids, GIDMap: gids}, nil
+		}
+		return nil, nil
+	}
+
+	// Ignore the annotation if not explicitly set in the config file.
+	if !s.config.AllowUsernsAnnotation {
+		return nil, nil
+	}
+
+	// expect a configuration like: private:uidmapping=0:1000:2000,2000:1000:2000;gidmapping=0:1000:4000,4000:1000:2000
+	parts := strings.SplitN(mode, ":", 2)
+
+	values := map[string]string{}
+	if len(parts) > 1 {
+		for _, r := range strings.Split(parts[1], ";") {
+			kv := strings.SplitN(r, "=", 2)
+			if len(kv) != 2 {
+				return nil, errors.Errorf("invalid argument: %q", r)
+			}
+			values[kv[0]] = kv[1]
+		}
+	}
+
+	_, uidMappingsPresent := values["uidmapping"]
+	_, gidMappingsPresent := values["uidmapping"]
+	// allow these options only if running as root
+	if uidMappingsPresent || gidMappingsPresent {
+		if sc.GetRunAsUser().Value != 0 {
+			return nil, errors.New("cannot use uidmapping or gidmapping if not running as root")
+		}
+	}
+
+	switch parts[0] {
+	case "auto":
+		const t = "true"
+		ret := &storage.IDMappingOptions{
+			AutoUserNs: true,
+		}
+		keepID := values["keep-id"] == t
```

Comment:
Could we add some comments here on the format for auto and what these values mean?